go/types/map_editor.go

```diff
@@ -0,0 +1,168 @@
+// Copyright 2017 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import (
+	"sort"
+
+	"github.com/attic-labs/noms/go/d"
+)
+
+// MapEditor allows for efficient editing of Map-typed prolly trees. Edits
+// are buffered to memory and can be applied via Build(), which returns a new
+// Map. Prior to Build(), Get() & Has() will return the value that the resulting
+// Map would return if it were built immediately prior to the respective call.
+// Note: The implementation biases performance towards a usage which applys
+// edits in key-order.
+type MapEditor struct {
+	m          Map
+	edits      mapEntrySlice // edits may contain duplicate key values, in which case, the last edit of a given key is used
+	normalized bool
+}
+
+func NewMapEditor(m Map) *MapEditor {
+	return &MapEditor{m, mapEntrySlice{}, true}
+}
+
+func (me *MapEditor) Build(vrw ValueReadWriter) Map {
+	if len(me.edits) == 0 {
+		return me.m // no edits
+	}
+
+	vr := me.m.sequence().valueReader()
+
+	me.normalize()
+
+	var ch *sequenceChunker
+	for i, kv := range me.edits {
+		if i+1 < len(me.edits) && me.edits[i+1].key.Equals(kv.key) {
```

Comment:
Did you consider keeping the edits sorted and only keeping the last of each entry? It seems like same amount of work for the computer, less memory, and reduce a lot of the code complexity.